pkg/chunk/encoding/doubledelta.go

```diff
@@ -133,26 +135,47 @@ func (c doubleDeltaEncodedChunk) Add(s model.SamplePair) ([]Chunk, error) {
 		}
 	}
 	if tb != ntb || vb != nvb || c.isInt() != nInt {
-		if len(c)*2 < cap(c) {
-			return transcodeAndAdd(newDoubleDeltaEncodedChunk(ntb, nvb, nInt, cap(c)), &c, s)
+		if len(*c)*2 < cap(*c) {
+			result, err := transcodeAndAdd(newDoubleDeltaEncodedChunk(ntb, nvb, nInt, cap(*c)), c, s)
+			if err != nil {
+				return nil, err
+			}
+			// We cannot handle >2 chunks returned as we can only return 1 chunk.
+			// Ideally there wont be >2 chunks, but if it happens to be >2,
+			// we fall through to perfom `addToOverflowChunk` instead.
+			if len(result) == 1 {
+				// Replace the current chunk with the new bigger chunk.
+				c0 := result[0].(*doubleDeltaEncodedChunk)
+				*c = *c0
+				return nil, nil
+			} else if len(result) == 2 {
+				// Replace the current chunk with the new bigger chunk
+				// and return the additional chunk.
+				c0 := result[0].(*doubleDeltaEncodedChunk)
+				c1 := result[1].(*doubleDeltaEncodedChunk)
+				*c = *c0
+				return c1, nil
+			}
 		}
+
 		// Chunk is already half full. Better create a new one and save the transcoding efforts.
-		return addToOverflowChunk(&c, s)
+		// We also perform this if `transcodeAndAdd` resulted in >2 chunks.
+		return addToOverflowChunk(s)
```

Comment:
This bit seems to be repeated four times - can we move it inside `transcodeAndAdd()` ?

Reply:
Given that there is typecast to specific types, some parts would again get repeated inside `transcodeAndAdd` for each type. Does that sound fine?

Reply:
OK leave it as it is.